src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidAccountManagerBroker.cs

```diff
@@ -0,0 +1,455 @@
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            AuthenticationContinuationHelper.LastRequestLogger = _logger;
+            _brokerHelper = new AndroidBrokerHelper(Application.Context, logger);
+        }
+
+        public bool IsBrokerInstalledAndInvokable()
+        {
+            return _brokerHelper.IsBrokerInstalledAndInvokable();
+        }
+
+        public async Task<MsalTokenResponse> AcquireTokenInteractiveAsync(
+            AuthenticationRequestParameters authenticationRequestParameters,
+            AcquireTokenInteractiveParameters acquireTokenInteractiveParameters)
+        {
+
+            CheckPowerOptimizationStatus();
+
+            AndroidBrokerStaticHelper.InteractiveBrokerTokenResponse = null;
+
+            BrokerRequest brokerRequest = BrokerRequest.FromInteractiveParameters(
+                authenticationRequestParameters, acquireTokenInteractiveParameters);
+
+            // There can only be 1 broker request at a time so keep track of the correlation id
+            s_correlationId = brokerRequest.CorrelationId;
+
+            try
+            {
+                await InitiateBrokerHandshakeAsync().ConfigureAwait(false);
+                await AcquireTokenInteractiveViaBrokerAsync(brokerRequest).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                _logger.ErrorPiiWithPrefix(ex, "Android broker interactive invocation failed. ");
+                _brokerHelper.HandleBrokerOperationError(ex);
+            }
+
+            using (_logger.LogBlockDuration("Waiting for Android broker response. "))
+            {
+                await AndroidBrokerStaticHelper.ReadyForResponse.WaitAsync().ConfigureAwait(false);
+                return AndroidBrokerStaticHelper.InteractiveBrokerTokenResponse;
+            }
+        }
+
+        public async Task<MsalTokenResponse> AcquireTokenSilentAsync(
+            AuthenticationRequestParameters authenticationRequestParameters,
+            AcquireTokenSilentParameters acquireTokenSilentParameters)
+        {
+            CheckPowerOptimizationStatus();
+
+            BrokerRequest brokerRequest = BrokerRequest.FromSilentParameters(
+                authenticationRequestParameters, acquireTokenSilentParameters);
+
+            try
+            {
+                await InitiateBrokerHandshakeAsync().ConfigureAwait(false);
+                var androidBrokerTokenResponse = await AcquireTokenSilentViaBrokerAsync(brokerRequest).ConfigureAwait(false);
+                return androidBrokerTokenResponse;
+            }
+            catch (Exception ex)
+            {
+                _logger.ErrorPiiWithPrefix(ex, "Android broker silent invocation failed. ");
+                _brokerHelper.HandleBrokerOperationError(ex);
+                throw;
+            }
+        }
+
+        private async Task AcquireTokenInteractiveViaBrokerAsync(BrokerRequest brokerRequest)
```

Comment:
>BrokerRequest brokerRequest [](start = 65, length = 27)

for this and the silent request, you might not need to pass in the request, just create it in the method, it's not used in the outer method anyway.

Reply:
If I create it here then I would need to pass in 2 variables to this function in order to create the brokerRequest: "(authenticationRequestParameters, acquireTokenSilentParameters)"

so I figured it was better to create the request initially and only pass it around.

Reply:
true. didn't see that. thanks.

---
In reply to: [570590223](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2380#discussion_r570590223) [](ancestors = 570590223)